src/Cli/dotnet/CommonOptions.cs

```diff
@@ -100,6 +116,74 @@ public static bool VerbosityIsDetailedOrDiagnostic(this VerbosityOptions verbosi
                 verbosity.Equals(VerbosityOptions.d) ||
                 verbosity.Equals(VerbosityOptions.detailed);
         }
+
+        internal static IEnumerable<string> ResolveArchOptionToRuntimeIdentifier(string arg, ParseResult parseResult)
+        {
+            if (parseResult.HasOption(RuntimeOption(string.Empty).Aliases.First()))
+            {
+                throw new GracefulException(CommonLocalizableStrings.CannotSpecifyBothRuntimeAndArchOptions);
+            }
+
+            if (parseResult.BothArchAndOsOptionsSpecified())
+            {
+                // ResolveOsOptionToRuntimeIdentifier handles resolving the RID when both arch and os are specified
+                return Array.Empty<string>();
+            }
+
+            return ResolveRidShorthandOptions(null, arg);
+        }
+
+        internal static IEnumerable<string> ResolveOsOptionToRuntimeIdentifier(string arg, ParseResult parseResult)
+        {
+            if (parseResult.HasOption(RuntimeOption(string.Empty).Aliases.First()))
+            {
+                throw new GracefulException(CommonLocalizableStrings.CannotSpecifyBothRuntimeAndOsOptions);
+            }
+
+            if (parseResult.BothArchAndOsOptionsSpecified())
+            {
+                return ResolveRidShorthandOptions(arg, parseResult.ValueForOption<string>(CommonOptions.ArchitectureOption().Aliases.First()));
+            }
+
+            return ResolveRidShorthandOptions(arg, null);
+        }
+
+        private static IEnumerable<string> ResolveRidShorthandOptions(string os, string arch)
+        {
+            var properties = new string[] { $"-property:RuntimeIdentifier={ResolveRidShorthandOptionsToRuntimeIdentifier(os, arch)}" };
+            return properties;
+        }
+
+        internal static string ResolveRidShorthandOptionsToRuntimeIdentifier(string os, string arch)
+        {
+            var currentRid = GetCurrentRuntimeId();
+            os = string.IsNullOrEmpty(os) ? GetOsFromRid(currentRid) : os;
+            arch = string.IsNullOrEmpty(arch) ? GetArchFromRid(currentRid) : arch;
+            return $"{os}-{arch}";
+        }
+
+        private static string GetCurrentRuntimeId()
+        {
+            var dotnetRootPath = Path.GetDirectoryName(Environment.ProcessPath);
+            // When running under test the path does not always contain "dotnet" and Product.Version is empty.
+            dotnetRootPath = dotnetRootPath.Contains("dotnet") ? dotnetRootPath : Path.Combine(dotnetRootPath, "dotnet");
```

Comment:
I still think that doing `path.Contains("dotnet")` is not the best. If I close the SDK repo to `F:\dotnet\sdk` (which is actually where I have it on my machine), that condition will be true for everything from the repo - including the `testhost` which comes from the repo-local SDK. But that doesn't mean that the path will point to `dotnet.exe`.

Reply:
Updated to check the file name in the latest commit.